passes/trees.nim

```diff
@@ -0,0 +1,173 @@
+## Implements the ``PackedTree`` storage type, for space efficient storage of
+## arbitrary trees.
+
+import
+  std/strutils,
+  experimental/sexp
+
+type
+  TreeNode*[T] = object
+    ## A single node. How the val field is interpreted depens on the kind. For
+    ## leaf nodes, the field's meaning is decided externally, while for non-
+    ## leaf nodes, `val` stores how many child nodes the node has.
+    kind*: T
+    val*: uint32
+
+  PackedTree*[T] = object
+    ## Stores a node tree packed together in a single sequence.
+    nodes: seq[TreeNode[T]]
+    numbers: seq[uint64]
+    # TODO: use a BiTable for the numbers
+
+  NodeIndex* = distinct uint32
+
+const
+  ExternalFlag = 0x8000_0000'u32
+
+proc `[]`*[T](t: PackedTree[T], at: NodeIndex): TreeNode[T] {.inline.} =
+  t.nodes[ord at]
+
+proc next*(t: PackedTree, i: NodeIndex): NodeIndex =
+  ## Returns the index of the first node following `i` that is not a child
+  ## node of `i`.
+  mixin isAtom
+  var
+    i = uint32(i)
+    last = i
+  while i <= last:
+    if not isAtom(t.nodes[i].kind):
+      last += t.nodes[i].val
+    inc i
+
+  result = NodeIndex(i)
+
+proc child*(t: PackedTree, i: NodeIndex, child: Natural): NodeIndex =
+  ## Returns the index of the `child`-th subnode of `i`.
+  var i = uint32(i) + 1
+  for j in 0..<child:
+    i = uint32 t.next(NodeIndex i)
+  result = NodeIndex i
+
+proc child*(t: PackedTree, i: NodeIndex, child: BackwardsIndex
+           ): NodeIndex {.inline.} =
+  ## Returns the index of the `child`-th (backwards) subnode of `i`.
+  t.child(i, t[i].val.int - ord(child))
+
+proc child*(t: PackedTree, i: Natural): NodeIndex =
+  ## Returns the index of the `i`th subnode of the top-level node.
+  child(t, NodeIndex(0), i)
+
+proc `[]`*[T](t: PackedTree[T], i: NodeIndex, child: Natural
+             ): TreeNode[T] {.inline.} =
+  t.nodes[ord t.child(i, child)]
+
+proc len*(t: PackedTree, i: NodeIndex): int =
+  t.nodes[ord i].val.int
+
+proc last*(tree: PackedTree, n: NodeIndex): NodeIndex =
+  tree.child(n, tree.len(n) - 1)
+
+iterator items*(t: PackedTree, at: NodeIndex): NodeIndex =
+  ## Returns the index of each child node of the node at `at`.
+  var n = NodeIndex(ord(at) + 1)
+  for i in 0..<t.len(at):
+    yield n
+    n = t.next(n)
+
+iterator items*(t: PackedTree, at: NodeIndex; start: int; last = ^1): NodeIndex =
+  ## Returns the index of each child node of the node at `at`, only
+  ## including subnode in slice ``start..last``.
+  var n = NodeIndex(ord(at) + 1)
```

Comment:
```suggestion
  assert start <= last, "cannot start after the last node"
  var n = NodeIndex(ord(at) + 1)
```

Reply:
I'm not sure. `items` works even if start > last, and being able to specify such empty ranges *could* be useful for reducing branching or arithmetic at the callsite.